File: day05/part1.py
import argparse
from typing import Callable
from typing import List

from support import timing

# 99 -> halt
# 1 -> read from 2 positions, add and store in third
#   (positions are directly after)
# 2 -> mult
# jump forward 4 after processing
"""\
Opcode 3 takes a single integer as input and saves it to the address given
by its only parameter. For example, the instruction 3,50 would take an input
value and store it at address 50.
Opcode 4 outputs the value of its only parameter. For example, the
instruction 4,50 would output the value at address 50.
"""


def run(prog: List[int], fn: Callable[[int], None]) -> List[int]:
    def parameter(instr: int, n: int) -> int:
        mode = instr // (10 ** (n + 1)) % 10
        if mode == 0:
            return prog[prog[pc + n]]
        elif mode == 1:
            return prog[pc + n]
        else:
            raise NotImplementedError(mode)

    pc = 0
    while pc < len(prog):
        instr = prog[pc]
        opc = instr % 100
        if opc == 99:
            return prog
        elif opc == 1:
            # can you assign in immediate mode?
            prog[prog[pc + 3]] = parameter(instr, 1) + parameter(instr, 2)
            pc += 4
        elif opc == 2:
            prog[prog[pc + 3]] = parameter(instr, 1) * parameter(instr, 2)
            pc += 4
        elif opc == 3:
            prog[prog[pc + 1]] = 1  # 1 is the hardcoded input
            pc += 2
        elif opc == 4:
            fn(parameter(instr, 1))
            pc += 2
        else:
            raise AssertionError(f'unreachable? {prog} {pc}')
    raise AssertionError(f'unreachable? {prog} {pc}')


def compute(s: str) -> None:
    prog = [int(part) for part in s.strip().split(',')]
    run(prog, print)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('data_file')
    args = parser.parse_args()

    with open(args.data_file) as f, timing():
        compute(f.read())

    return 0


if __name__ == '__main__':
    exit(main())

File: day05/part1.js
'use strict';

const fs = require('fs');

if (process.argv.length !== 3) {
    console.error(`usage: node ${process.argv[1]} FILENAME`);
    process.exit(1)
}


function compute(prog, fn) {
    function parameter(instr, n) {
        const mode = Math.trunc(instr / (10 ** (n + 1))) % 10;
        switch (mode) {
            case 0:
                return prog[prog[pc + n]];
            case 1:
                return prog[pc + n];
            default:
                throw new Error('OHNOSE!!!');
        }
    }

    let pc = 0;
    for (;;) {
        const instr = prog[pc];
        const opc = instr % 100;
        switch (opc) {
            case 99:
                return prog[0];
            case 1:
                prog[prog[pc + 3]] = parameter(instr, 1) + parameter(instr, 2);
                pc += 4;
                break;
            case 2:
                prog[prog[pc + 3]] = parameter(instr, 1) * parameter(instr, 2);
                pc += 4;
                break;
            case 3:
                prog[prog[pc + 1]] = 1;  // hardcoded value from puzzle
                pc += 2;
                break;
            case 4:
                fn(parameter(instr, 1));
                pc += 2;
                break;
            default:
                throw new Error('OHNOES!');
        }
    }
}

const prog =
    fs.readFileSync(process.argv[2], {encoding: 'UTF-8'})
        .trim()
        .split(/,/)
        .map((s) => parseInt(s, 10));

compute(prog, console.log);
